doc/neps/nep-0041-improved-dtype-support.rst

```diff
@@ -0,0 +1,706 @@
+fully featured datatypes and reorganization of the NumPy datatype system.
+Phase I concerns mainly private APIs, while phase II is expected to include
+a slow creation of a *preliminary* public API which will develop into the
+final stable API.
+
+Some of the benefits of a large refactor may only become evident after the full
+deprecation of the current legacy implementation (i.e. larger code removals).
+However, these steps are necessary for improvements to many parts of the
+core NumPy API, and are expected to make the implementation generally
+easier to understand.
+Pushing forward with Phase I cleans up the concepts of the API to coincide
+with how Python is typically organized and unlocks the ability to improve
+parts of NumPy which currently are not extensible.
+
+The following picture gives another overview of the overall design idea.
+Note that this NEP only regards Phase I (shaded area),
+the rest encompasses Phase II and the design choices are up for discussion,
+however, it highlights that the DType datatype class is the central, necessary
+concept:
+
+.. image:: _static/nep-0041-mindmap.svg
+
+
+First steps directly related to this NEP
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The required changes necessary to NumPy are massive and touch a large part
+of the code base.
+We thus propose the implementation of the above steps listed in Phases I and II.
+
+Although it is possible that new DTypes will only be useful after Phase II
+is finished, this NEP proposes to start implementation in an incremental way.
+This means that in a first step the ``DType`` classes will be added, with
+all, or most, new exposed API points giving a ``PreliminaryDTypeAPIWarning``.
+
+This allows for smaller patches and further future changes. In these first
+steps, only a very limited new C-API will be exposed to allow writing tests,
+with no expectation that early adopters will begin using it.
+The addition of ``DType`` will then enable addressing other changes
+more incrementally:
+
+1. New machinery for array coercion, with the goal of enabling user DTypes
+   with appropriate class methods.
+2. The replacement or wrapping of the current casting machinery.
+3. Incremental redefinition of the current ``ArrFunctions`` slots into
+   DType method slots.
+
+Parallel to these, after step 1. is finished, Phase II of revising the
+UFunc machinery can be started.
+
+In particular the step of creating a C defined ``DTypeMeta`` class with its
+instances being ``DTypeClasses`` is a necessary first step to add the ability
+of storing custom slots on the datatype in C.
+This ``DTypeMeta`` must thus be implemented before being widely used to
+restructure or enhance current code, thus we propose to proceed with mainly
+private additions to the DType classes.
+
+
+Backward compatibility
+----------------------
+
+While the actual backward compatibility impact is not yet fully clear,
+we anticipate, and accept the following changes:
```

Comment:
Are these backward-compatibility considerations related only to Phase I (i.e. NEP 41-related) changes? Or is this an accounting of changes that are expected to result from the entire refactor? If the former, you should explicitly state that.

Reply:
Made it explicitly Phase I and II, since I personally think it is not useful to only list Phase I impact.